### Assignment 1 : Non personalized and stereotype-baed recommenders
* data : movie ID 와 user ID 두 축으로 이루어진 rating data

In [1]:
import numpy as np
import pandas as pd

In [101]:
df = pd.read_csv("HW1-data.csv")

In [102]:
df=df.set_index("User")

In [103]:
rating = df.drop(["Gender (1 =F, 0=M)"],axis=1)

In [104]:
rating

260: Star Wars: Episode IV - A New Hope (1977)  \
User                                                   
755                                              1.0   
5277                                             5.0   
1577                                             NaN   
4388                                             NaN   
1202                                             4.0   
3823                                             2.0   
5448                                             NaN   
5347                                             4.0   
4117                                             5.0   
2765                                             4.0   
5450                                             2.0   
139                                              3.0   
1940                                             2.0   
3118                                             3.0   
4656                                             4.0   
4796                                             NaN   
6037                                             NaN   
3048                                             4.0   
4790                                             5.0   
4489                                             1.0   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
User                                                            
755                                                 5.0         
5277                                                3.0         
1577                                                NaN         
4388                                                3.0         
1202                                                3.0         
3823                                                4.0         
5448                                                NaN         
5347                                                NaN         
4117                                                1.0         
2765                                                2.0         
5450                                                1.0         
139                                                 5.0         
1940                                                3.0         
3118                                                NaN         
4656                                                4.0         
4796                                                NaN         
6037                                                NaN         
3048                                                5.0         
4790                                                1.0         
4489                                                2.0         

      356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
User                                                                    
755                        2.0                                    NaN   
5277                       NaN                                    2.0   
1577                       NaN                                    5.0   
4388                       NaN                                    NaN   
1202                       4.0                                    1.0   
3823                       4.0                                    4.0   
5448                       3.0                                    1.0   
5347                       NaN                                    NaN   
4117                       NaN                                    4.0   
2765                       NaN                                    5.0   
5450                       5.0                                    NaN   
139                        2.0                                    NaN   
1940                       NaN                                    5.0   
3118                       3.0                                    NaN   
4656                       NaN                                    NaN   
4796                       1.0                                    NaN   
6037                       N

1. Mean Rating: Calculate the mean rating for each movie, order with the highest rating listed first, and submit the top three (along with the mean scores for the top two).
* 평균 평점 계산하기

In [100]:
mean_rating = np.mean(rating,axis=0)
mean_rating.sort_values(ascending=False)

318: Shawshank Redemption, The (1994)                      3.600000
260: Star Wars: Episode IV - A New Hope (1977)             3.266667
541: Blade Runner (1982)                                   3.222222
1265: Groundhog Day (1993)                                 3.166667
593: Silence of the Lambs, The (1991)                      3.062500
296: Pulp Fiction (1994)                                   3.000000
527: Schindler's List (1993)                               3.000000
1210: Star Wars: Episode VI - Return of the Jedi (1983)    3.000000
34: Babe (1995)                                            3.000000
2028: Saving Private Ryan (1998)                           3.000000
3578: Gladiator (2000)                                     2.916667
1198: Raiders of the Lost Ark (1981)                       2.909091
2396: Shakespeare in Love (1998)                           2.909091
2571: Matrix, The (1999)                                   2.833333
2762: Sixth Sense, The (1999)                   

2. Rating Count (popularity): Count the number of ratings for each movie, order with the most number of ratings first, and submit the top three (along with the counts for the top two). 
* 평점 개수 세기 (인기의 지표) 

In [121]:
count = rating.count()
count.sort_values(ascending=False)

1: Toy Story (1995)                                        17
593: Silence of the Lambs, The (1991)                      16
260: Star Wars: Episode IV - A New Hope (1977)             15
1210: Star Wars: Episode VI - Return of the Jedi (1983)    14
780: Independence Day (ID4) (1996)                         13
2762: Sixth Sense, The (1999)                              12
3578: Gladiator (2000)                                     12
527: Schindler's List (1993)                               12
2571: Matrix, The (1999)                                   12
1265: Groundhog Day (1993)                                 12
2916: Total Recall (1990)                                  12
1198: Raiders of the Lost Ark (1981)                       11
2396: Shakespeare in Love (1998)                           11
1259: Stand by Me (1986)                                   11
296: Pulp Fiction (1994)                                   11
2028: Saving Private Ryan (1998)                           11
318: Sha

3. % of ratings 4+ (liking): Calculate the percentage of ratings for each movie that are 4 or higher. Order with the highest percentage first, and submit the top three (along with the percentage for the top two).  Notice that the three different measures of "best" reflect different priorities and give different results; this should help you see why you need to be thoughtful about what metrics you use. 

In [115]:
rating >= 4

260: Star Wars: Episode IV - A New Hope (1977)  \
User                                                   
755                                            False   
5277                                            True   
1577                                           False   
4388                                           False   
1202                                            True   
3823                                           False   
5448                                           False   
5347                                            True   
4117                                            True   
2765                                            True   
5450                                           False   
139                                            False   
1940                                           False   
3118                                           False   
4656                                            True   
4796                                           False   
6037                                           False   
3048                                            True   
4790                                            True   
4489                                           False   

      1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
User                                                            
755                                                True         
5277                                              False         
1577                                              False         
4388                                              False         
1202                                              False         
3823                                               True         
5448                                              False         
5347                                              False         
4117                                              False         
2765                                              False         
5450                                              False         
139                                                True         
1940                                              False         
3118                                              False         
4656                                               True         
4796                                              False         
6037                                              False         
3048                                               True         
4790                                              False         
4489                                              False         

      356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
User                                                                    
755                      False                                  False   
5277                     False                                  False   
1577                     False                                   True   
4388                     False                                  False   
1202                      True                                  False   
3823                      True                                   True   
5448                     False                                  False   
5347                     False                                  False   
4117                     False                                   True   
2765                     False                                   True   
5450                      True                                  False   
139                      False                                  False   
1940                     False                                   True   
3118                     False                                  False   
4656                     False                                  False   
4796                     False                                  False   
6037                     Fal

In [124]:
positive_count = np.sum(rating >= 4,axis=0)

In [133]:
perc_pos_count = positive_count/count*100
perc_pos_count.sort_values(ascending=False)

318: Shawshank Redemption, The (1994)                      70.000000
260: Star Wars: Episode IV - A New Hope (1977)             53.333333
3578: Gladiator (2000)                                     50.000000
541: Blade Runner (1982)                                   44.444444
593: Silence of the Lambs, The (1991)                      43.750000
2571: Matrix, The (1999)                                   41.666667
1265: Groundhog Day (1993)                                 41.666667
34: Babe (1995)                                            40.000000
296: Pulp Fiction (1994)                                   36.363636
1259: Stand by Me (1986)                                   36.363636
2028: Saving Private Ryan (1998)                           36.363636
1210: Star Wars: Episode VI - Return of the Jedi (1983)    35.714286
1: Toy Story (1995)                                        35.294118
527: Schindler's List (1993)                               33.333333
2762: Sixth Sense, The (1999)     

4. Top movies for someone who has seen Toy Story: Calculate movies that most often occur with Movie #1:  Toy Story, using the (x+y)/x method described in class. In other words, for each movie, *calculate the percentage of Toy Story raters who also rated that movie*. Order with the highest percentage first, and submit the top 3 (along with the correlations for the top two).  Note, you will have ties -- to break the ties, use the lowest-numbered movie as the higher-ranked one.  In other words, if Movies 541 and 318 are tied, then 318 gets the higher rank.
* Product Assocition (이 제품을 구매한 사람이 같이 구매한 제품) 
* P(A|B) 계산
* B = toy story

In [153]:
pa = rating.isnull() == False
toy = pa["1: Toy Story (1995)"]
pa = pa.drop(["1: Toy Story (1995)"],axis=1)

In [163]:
# 해당 영화와 toy story를 둘다 본 사람들의 수 
for i in range(19):
    pa.iloc[:,i] = pa.iloc[:,i] * toy
np.sum(pa,axis=0)

C:\Users\User\anaconda3\lib\site-packages\pandas\core\computation\expressions.py:204: UserWarning: evaluating in Python space because the '*' operator is not supported by numexpr for the bool dtype, use '&' instead
  warnings.warn(


260: Star Wars: Episode IV - A New Hope (1977)             14
1210: Star Wars: Episode VI - Return of the Jedi (1983)    13
356: Forrest Gump (1994)                                    7
318: Shawshank Redemption, The (1994)                       8
593: Silence of the Lambs, The (1991)                      13
3578: Gladiator (2000)                                      9
2028: Saving Private Ryan (1998)                            8
296: Pulp Fiction (1994)                                   10
1259: Stand by Me (1986)                                    9
2396: Shakespeare in Love (1998)                            8
2916: Total Recall (1990)                                  11
780: Independence Day (ID4) (1996)                         13
541: Blade Runner (1982)                                    7
1265: Groundhog Day (1993)                                 11
2571: Matrix, The (1999)                                    9
527: Schindler's List (1993)                               10
2762: Si

In [174]:
# product assication (P(movies|toy story))
toy_association= np.sum(pa,axis=0)/17
toy_association.sort_values(ascending=False)

260: Star Wars: Episode IV - A New Hope (1977)             0.823529
593: Silence of the Lambs, The (1991)                      0.764706
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.764706
780: Independence Day (ID4) (1996)                         0.764706
1265: Groundhog Day (1993)                                 0.647059
2916: Total Recall (1990)                                  0.647059
2762: Sixth Sense, The (1999)                              0.588235
527: Schindler's List (1993)                               0.588235
296: Pulp Fiction (1994)                                   0.588235
1198: Raiders of the Lost Ark (1981)                       0.529412
3578: Gladiator (2000)                                     0.529412
1259: Stand by Me (1986)                                   0.529412
2571: Matrix, The (1999)                                   0.529412
2396: Shakespeare in Love (1998)                           0.470588
2028: Saving Private Ryan (1998)                

5. Correlation with Toy Story:  Calculate the correlation between the vectors of ratings for Toy Story and each other movie.   Order by the highest correlation (positive only) and submit the top 3 along with the correlation values for the top 2.  Notice the differences between co-occurrence and correlation; these metrics are showing different types of relationships.

In [180]:
toy_corr =rating.corr()["1: Toy Story (1995)"]
toy_corr.sort_values(ascending=False)

1: Toy Story (1995)                                        1.000000
318: Shawshank Redemption, The (1994)                      0.888523
34: Babe (1995)                                            0.811107
296: Pulp Fiction (1994)                                   0.709842
2028: Saving Private Ryan (1998)                           0.596849
356: Forrest Gump (1994)                                   0.522913
541: Blade Runner (1982)                                   0.329634
3578: Gladiator (2000)                                     0.253396
2916: Total Recall (1990)                                  0.163796
2396: Shakespeare in Love (1998)                           0.101768
1265: Groundhog Day (1993)                                -0.062858
780: Independence Day (ID4) (1996)                        -0.069923
260: Star Wars: Episode IV - A New Hope (1977)            -0.119005
527: Schindler's List (1993)                              -0.220315
2762: Sixth Sense, The (1999)                   

6. Mean rating difference by gender:  First, recompute the mean rating for each movie separately for males and for females.  And calculate the overall mean rating (across all ratings) for males and females. Submit the two movies that have the greatest differences (one where men are most above women, and one where women are most above men) along with the differences in average.  Also submit the difference in overall rating averages (female average - male average).

In [189]:
gender_mean = df.groupby("Gender (1 =F, 0=M)").agg("mean")

In [196]:
gender_mean["all movie"] = np.mean(gender_mean,axis=1)
gender_mean.loc["diff"] = gender_mean.loc[0] - gender_mean.loc[1]
gender_mean.loc["diff"].sort_values(ascending=False)

1198: Raiders of the Lost Ark (1981)                       1.666667
527: Schindler's List (1993)                               1.000000
2571: Matrix, The (1999)                                   0.742857
1265: Groundhog Day (1993)                                 0.666667
593: Silence of the Lambs, The (1991)                      0.619048
1259: Stand by Me (1986)                                   0.571429
2916: Total Recall (1990)                                  0.485714
2028: Saving Private Ryan (1998)                           0.392857
780: Independence Day (ID4) (1996)                         0.190476
1210: Star Wars: Episode VI - Return of the Jedi (1983)    0.000000
all movie                                                 -0.115000
3578: Gladiator (2000)                                    -0.166667
260: Star Wars: Episode IV - A New Hope (1977)            -0.303571
2762: Sixth Sense, The (1999)                             -0.333333
318: Shawshank Redemption, The (1994)           

7. Next, compute the % of ratings 4+ separately for males and females.  You'll be asked to submit two movies as above (largest difference in each direction). And again you'll indicate whether men or women are more likely to rate movies 4 stars or above.

In [245]:
gender_pos_count = pd.concat([df.iloc[:,0],df.iloc[:,1:] >= 4],axis=1).groupby("Gender (1 =F, 0=M)").agg("sum")

In [246]:
gender_count= pd.concat([df.iloc[:,0],df.iloc[:,1:]],axis=1).groupby("Gender (1 =F, 0=M)").agg("count")

In [251]:
gender_pos

260: Star Wars: Episode IV - A New Hope (1977)  \
Gender (1 =F, 0=M)                                                   
0                                                         0.500000   
1                                                         0.571429   
diff                                                     -0.071429   

                    1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
Gender (1 =F, 0=M)                                                            
0                                                                0.25         
1                                                                0.50         
diff                                                            -0.25         

                    356: Forrest Gump (1994)  \
Gender (1 =F, 0=M)                             
0                                        0.0   
1                                        0.5   
diff                                    -0.5   

                    318: Shawshank Redemption, The (1994)  \
Gender (1 =F, 0=M)                                          
0                                                     0.6   
1                                                     0.8   
diff                                                 -0.2   

                    593: Silence of the Lambs, The (1991)  \
Gender (1 =F, 0=M)                                          
0                                                0.555556   
1                                                0.285714   
diff                                             0.269841   

                    3578: Gladiator (2000)  1: Toy Story (1995)  \
Gender (1 =F, 0=M)                                                
0                                      0.5             0.200000   
1                                      0.5             0.571429   
diff                                   0.0            -0.371429   

                    2028: Saving Private Ryan (1998)  \
Gender (1 =F, 0=M)                                     
0                                           0.285714   
1                                           0.500000   
diff                                       -0.214286   

                    296: Pulp Fiction (1994)  1259: Stand by Me (1986)  ...  \
Gender (1 =F, 0=M)                                                      ...   
0                                   0.250000                  0.250000  ...   
1                                   0.666667                  0.428571  ...   
diff                               -0.416667                 -0.178571  ...   

                    2916: Total Recall (1990)  \
Gender (1 =F, 0=M)                              
0                                         0.2   
1                                         0.0   
diff                                      0.2   

                    780: Independence Day (ID4) (1996)  \
Gender (1 =F, 0=M)                                       
0                                             0.285714   
1                                             0.333333   
diff                                         -0.047619   

                    541: Blade Runner (1982)  1265: Groundhog Day (1993)  \
Gender (1 =F, 0=M)                                                         
0                                       0.20                    0.500000   
1                                       0.75                    0.333333   
diff                                   -0.55                    0.166667   

                    2571: Matrix, The (1999)  527: Schindler's List (1993)  \
Gender (1 =F, 0=M)                                                           
0                                   0.571429                      0.500000   
1                                   0.200000                      0.166667   
diff                                0.371429                      0.333333   

                    2762: Sixth Sense, The (1999)  \
Gender (1 =F, 0=M)                           

In [250]:
gender_pos = gender_pos_count/gender_count
gender_pos.loc["diff"] = gender_pos.loc[0] - gender_pos.loc[1]
gender_pos.loc["diff"].sort_values(ascending=False)

1198: Raiders of the Lost Ark (1981)                       0.500000
2571: Matrix, The (1999)                                   0.371429
527: Schindler's List (1993)                               0.333333
593: Silence of the Lambs, The (1991)                      0.269841
2916: Total Recall (1990)                                  0.200000
1265: Groundhog Day (1993)                                 0.166667
2762: Sixth Sense, The (1999)                              0.000000
3578: Gladiator (2000)                                     0.000000
780: Independence Day (ID4) (1996)                        -0.047619
260: Star Wars: Episode IV - A New Hope (1977)            -0.071429
all movie                                                 -0.114008
1259: Stand by Me (1986)                                  -0.178571
318: Shawshank Redemption, The (1994)                     -0.200000
2028: Saving Private Ryan (1998)                          -0.214286
1210: Star Wars: Episode VI - Return of the Jedi

In [266]:
np.sum(gender_pos_count,axis=1)/np.sum(gender_count,axis=1)

Gender (1 =F, 0=M)
0    0.338583
1    0.421053
dtype: float64

In [267]:
0.421053-0.338583

0.08246999999999999